stdlib/SuiteSparse/src/umfpack.jl

```diff
@@ -176,6 +176,65 @@ lu(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}}};
     "dense LU.")))
 lu(A::SparseMatrixCSC; check::Bool = true) = lu(float(A); check = check)
 
+"""
+    lu!(F::UmfpackLU, A::SparseMatrixCSC; check = true) -> F::UmfpackLU
+
+Compute the LU factorization of a sparse matrix `A`, reusing the symbolic
+factorization of an already existing LU factorization stored in `F`. The
+sparse matrix `A` must have an identical nonzero pattern as the matrix used
+to create the LU factorization `F`, otherwise an error is thrown.
+
+When `check = true`, an error is thrown if the decomposition fails.
+When `check = false`, responsibility for checking the decomposition's
+validity (via [`issuccess`](@ref)) lies with the user.
+
+!!! note
+    `lu!(F::UmfpackLU, A::SparseMatrixCSC)` uses the UMFPACK library that is part of
+    SuiteSparse. As this library only supports sparse matrices with [`Float64`](@ref) or
+    `ComplexF64` elements, `lu!` converts `A` into a copy that is of type
+    `SparseMatrixCSC{Float64}` or `SparseMatrixCSC{ComplexF64}` as appropriate.
+
+!!! compat "Julia 1.4"
+    `lu!` for `UmfpackLU` requires at least Julia 1.4.
+
+# Examples
+```jldoctest
+julia> A = sparse(Float64[1.0 2.0; 0.0 3.0]);
+
+julia> F = lu(A);
+
+julia> B = sparse(Float64[1.0 1.0; 0.0 1.0]);
+
+julia> lu!(F, B);
+
+julia> F \\ ones(2)
+2-element Array{Float64,1}:
+ 0.0
+ 1.0
+```
+"""
+function lu!(F::UmfpackLU, S::SparseMatrixCSC{<:UMFVTypes,<:UMFITypes}; check::Bool = true)
+    zerobased = getcolptr(S)[1] == 0
+    F.m = size(S, 1)
+    F.n = size(S, 2)
+    F.colptr = zerobased ? copy(getcolptr(S)) : decrement(getcolptr(S))
+    F.rowval = zerobased ? copy(rowvals(S)) : decrement(rowvals(S))
+    F.nzval = copy(nonzeros(S))
```

Comment:
These copy instructions are certainly quite expensive.
I was wondering if there is a way to prevent these?

Reply:
I don't see any way to prevent it. We do the same in `lu` above, and can't write into the fields of `S`.

Reply:
Yes, this was just a thought that came to my mind, if it was technically possible to restructure `umfpack.jl`, such that UMFPACK has direct access to the data and copying the data can be avoided. I think it's not a big deal if we perform a LU decomposition once, but it probably has a big impact for repeatedly performed LU factorizations.

Reply:
Nevertheless, your PR is a strict improvement, because we copy as in `lu`, but save the symbolic factorization step, right? I'm trying to understand if this issue is merge-blocking, or if this is something for future development and the PR can be merged as is.

Reply:
Yes exactly, this PR allows the user to reuse the symbolic factorization for repeatedly performed LU decompositions.

My comment above is out-of-scope for this PR, but might be interesting to think about for future development.